clippy_lints/src/utils/internal_lints.rs

```diff
@@ -391,3 +402,188 @@ fn is_trigger_fn(fn_kind: FnKind<'_>) -> bool {
+
+impl  EarlyLintPass for CollapsibleCalls {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &AstExpr) {
+        use ast::{StmtKind, ExprKind};
+
+        if_chain! {
+            if let ExprKind::Call(ref func, ref and_then_args) = expr.kind;
+            if let ExprKind::Path(None, ref path) = func.kind;
+            if match_path_ast(path, &["span_lint_and_then"]);
+            if and_then_args.len() == 5;
+            if let ExprKind::Closure(_, _, _, _, block, _) = &and_then_args[4].kind;
+            if let ExprKind::Block(block, _) = &block.kind;
+            let stmts = &block.stmts;
+            if stmts.len() == 1;
+            if let StmtKind::Semi(only_expr) = &stmts[0].kind;
+            if let ExprKind::MethodCall(ref ps, ref span_call_args) = &only_expr.kind;
+            let and_then_args = get_and_then_args(cx, and_then_args);
+            then {
+                match &*ps.ident.as_str() {
+                    "span_suggestion" =>
+                        suggest_span_suggestion(cx, expr, and_then_args, suggestion_args(cx, span_call_args)),
+                    "span_help" =>
+                        suggest_span_help(cx, expr, and_then_args, help_args(cx, span_call_args)),
+                    "span_note" =>
+                        suggest_span_note(cx, expr, and_then_args, note_args(cx, span_call_args)),
+                    _  => (),
+                }
+            }
+        }
+    }
+}
+
+struct AndThenArgs<'a> {
+    cx: Cow<'a, str>,
+    lint: Cow<'a, str>,
+    span: Cow<'a, str>,
+    msg: Cow<'a, str>,
+}
+
+fn get_and_then_args<'a>(cx: &EarlyContext<'_>, and_then_args: &Vec<P<AstExpr>>) -> AndThenArgs<'a>{
+    let cx_snippet = snippet(cx, and_then_args[0].span);
+    let lint_snippet= snippet(cx, and_then_args[1].span);
+    let span_snippet = snippet(cx, and_then_args[2].span);
+    let msg_snippet= snippet(cx, and_then_args[3].span);
+
+    AndThenArgs {
+        cx: cx_snippet,
+        lint: lint_snippet,
+        span: span_snippet,
+        msg: msg_snippet,
+    }
+}
+
+struct SuggestionArgs<'a> {
+    span: Cow<'a, str>,
+    help: Cow<'a, str>,
+    sugg: Cow<'a, str>,
+    applicability: Cow<'a, str>,
+}
+
+fn suggestion_args<'a>(cx: &EarlyContext<'_>, span_call_args: &Vec<P<AstExpr>>) -> SuggestionArgs<'a> {
+    let span_snippet_of_span_call = snippet(cx, span_call_args[0].span);
+    let help_snippet = snippet(cx, span_call_args[1].span);
+    let sugg_snippet = snippet(cx, span_call_args[2].span);
+    let applicability_snippet = snippet(cx, span_call_args[3].span);
+
+    SuggestionArgs {
+        span: span_snippet_of_span_call,
+        help: help_snippet,
+        sugg: sugg_snippet,
+        applicability: applicability_snippet,
+    }
+}
+
+fn suggest_span_suggestion(cx: &EarlyContext<'_>, expr: &AstExpr, and_then_args: AndThenArgs<'_>, suggestion_args: SuggestionArgs<'_>) {
+    if and_then_args.span == suggestion_args.span {
```

Comment:
Why is this condition test out to be false? I read the docs, they are both Cow<'a, str>, they should be equal.

Reply:
I tested out with adding and else branch and the else branch was run.

Reply:
You shouldn't determine if two things are the same, by comparing there snippets, but rather use `utils::hir_utils::SpanlessEq` to compare expressions. That should also simplify your code.